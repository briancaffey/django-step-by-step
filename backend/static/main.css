ul {
    padding-bottom: 0px !important;
}

.registration-form,
.login-form {
    width: 100%;
    max-width: 330px;
    padding: 15px;
    margin: auto;
}

.post-image {
    max-height: 200px;
    margin: auto;
    padding: 15px;
}

.fieldWrapper {
    padding-bottom: 15px;
}