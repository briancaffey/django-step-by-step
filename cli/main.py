import argparse
import subprocess

import boto3

def parse_args():
    parser = argparse.ArgumentParser(description="")

    parser.add_argument(
        "--base_env_name",
        type=str,
        required=True,
        default=None,
        help="Path to the R1 file"
    )

    parser.add_argument(
        "--app_env_name",
        type=str,
        required=True,
        default=None,
        help="Number of threads to use"
    )

    args = parser.parse_args()
    return args


def main():
    """
    This command is used to get access to a running container with ecs-exec

    This command should be IaC agnostic, so instead of using stack outputs
    it will use tags to get resources

    Items to look up via tags

    * cluster name
    * task id

    arguments:

    * base_env_name
    * app_env_name
    """

    args = parse_args()

    # use args to look up ecs cluster name
    # base_env_name = args.base_env_name
    app_env_name = args.app_env_name

    # use aws cli get the ecs cluster name with tags app-env = app_env_name and base-env = base_env_name
    cluster_name = f"{app_env_name}-cluster"

    ecs_client = boto3.client("ecs", region_name="us-east-1")
    task_id = ecs_client.list_tasks(cluster=cluster_name, serviceName=f"{app_env_name}-gunicorn")["taskArns"][0]

    print(f"Task ID: {task_id}")

    aws_command = [
        'aws', 'ecs', 'execute-command',
        '--cluster', cluster_name,
        '--task', task_id,
        '--command', 'bash',
        '--interactive', '--region', 'us-east-1'
    ]

    try:
        subprocess.run(aws_command, capture_output=False, text=True)
    except subprocess.CalledProcessError as e:
        print(f"Error executing command: {e}")


if __name__ == "__main__":
    main()
