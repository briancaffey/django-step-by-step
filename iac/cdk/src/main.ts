import { App, Stack, Tags } from 'aws-cdk-lib';
import { AdHocBase, AdHocApp } from 'cdk-django';

// look up the following values from environment variables
const certificateArn = process.env.ACM_CERTIFICATE_ARN || 'arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123456789012';
const domainName = process.env.DOMAIN_NAME || 'example.com';

// the environments for the base stack and the app stack
// the ad hoc environment for the following would have a URL of
// alpha.dev.domain.com
const adHocBaseEnvName = process.env.AD_HOC_BASE_NAME || 'dev';
const adHocAppEnvName = process.env.AD_HOC_APP_NAME || 'alpha';

const app = new App();

const env = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const adHocBaseStack = new Stack(app, 'AdHocBaseStack', { stackName: adHocBaseEnvName, env });
const adHocAppStack = new Stack(app, 'AdHocAppStack', { stackName: adHocAppEnvName, env });

const adHocBase = new AdHocBase(adHocBaseStack, 'AdHocBase', { certificateArn, domainName });

const adHocApp = new AdHocApp(adHocAppStack, 'AdHocApp', {
  baseStackName: adHocBaseEnvName,
  vpc: adHocBase.vpc,
  alb: adHocBase.alb,
  appSecurityGroup: adHocBase.appSecurityGroup,
  serviceDiscoveryNamespace: adHocBase.serviceDiscoveryNamespace,
  rdsInstance: adHocBase.databaseInstance,
  assetsBucket: adHocBase.assetsBucket,
  domainName: adHocBase.domainName,
  listener: adHocBase.listener,

  djangoSettingsModule: process.env.DJANGO_SETTINGS_MODULE || 'backend.settings.aws',
  backendVersion: process.env.BACKEND_VERSION || 'latest',
  frontendVersion: process.env.FRONTEND_VERSION || 'latest',
});

/**
 * Add tagging for this construct and all child constructs
 */
Tags.of(adHocBase).add('env', adHocBaseStack.stackName);

Tags.of(adHocApp).add('env', adHocAppStack.stackName);


app.synth();
