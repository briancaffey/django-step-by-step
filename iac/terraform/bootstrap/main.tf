terraform {
  required_version = ">=1.3.6"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "4.48.0"
    }
  }
}

provider "aws" {
  region = var.region
}

# S3 bucket for storing state file
resource "aws_s3_bucket" "this" {
  bucket = "${var.backend_name}-bucket"

  lifecycle {
    prevent_destroy = true
  }
}

resource "aws_s3_bucket_versioning" "this" {
  bucket = aws_s3_bucket.this.id
  versioning_configuration {
    status = "Enabled"
  }
}

# DynamoDB table for locking the state file
resource "aws_dynamodb_table" "this" {
  name           = "${var.backend_name}-lock-table"
  hash_key       = "LockID"
  read_capacity  = 20
  write_capacity = 20

  attribute {
    name = "LockID"
    type = "S"
  }

  lifecycle {
    prevent_destroy = true
  }
}

# ECR repository for the backend web app
resource "aws_ecr_repository" "backend" {
  name                 = "backend"
  image_tag_mutability = "MUTABLE"

  image_scanning_configuration {
    scan_on_push = true
  }
}

# ECR repository for the backend web app sidecar NGINX container
resource "aws_ecr_repository" "backend-nginx" {
  name                 = "backend-nginx"
  image_tag_mutability = "MUTABLE"

  image_scanning_configuration {
    scan_on_push = true
  }
}

# ECR repository for the frontend web app
resource "aws_ecr_repository" "frontend" {
  name                 = "frontend"
  image_tag_mutability = "MUTABLE"

  image_scanning_configuration {
    scan_on_push = true
  }
}
