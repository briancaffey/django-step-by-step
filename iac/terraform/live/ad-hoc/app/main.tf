terraform {
  required_version = ">=1.3.6"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "4.48.0"
    }
  }

  backend "s3" {}
}

provider "aws" {
  region = var.region
  default_tags {
    tags = {
      env                = terraform.workspace
      ad_hoc_env         = terraform.workspace
      ad_hoc_environment = "true"
    }
  }
}

# shared resources
# see https://github.com/briancaffey/terraform-aws-ad-hoc-environments

# https://www.terraform.io/language/settings/backends/s3#s3-state-storage
data "terraform_remote_state" "shared" {
  backend = "s3"
  config = {
    bucket               = var.s3_bucket
    key                  = "terraform.tfstate"
    region               = var.region
    workspace_key_prefix = "shared-resources"
  }
  workspace = var.shared_resources_workspace
}

# main

module "main" {
  source = "git::https://github.com/briancaffey/terraform-aws-django.git//modules/ad-hoc/app"

  # shared resources -- taken from terraform_remote_state data source above

  vpc_id                         = data.terraform_remote_state.shared.outputs.vpc_id
  private_subnets                = data.terraform_remote_state.shared.outputs.private_subnets
  public_subnets                 = data.terraform_remote_state.shared.outputs.public_subnets
  listener_arn                   = data.terraform_remote_state.shared.outputs.listener_arn
  alb_dns_name                   = data.terraform_remote_state.shared.outputs.alb_dns_name
  service_discovery_namespace_id = data.terraform_remote_state.shared.outputs.service_discovery_namespace_id
  task_role_arn                  = data.terraform_remote_state.shared.outputs.task_role_arn
  execution_role_arn             = data.terraform_remote_state.shared.outputs.execution_role_arn
  rds_address                    = data.terraform_remote_state.shared.outputs.rds_address
  alb_default_tg_arn             = data.terraform_remote_state.shared.outputs.alb_default_tg_arn
  ecs_sg_id                      = data.terraform_remote_state.shared.outputs.ecs_sg_id

  shared_resources_workspace = var.shared_resources_workspace

  # per environment settings -- taken from .tfvars files

  ecr_be_repo_url = var.ecr_be_repo_url
  ecr_fe_repo_url = var.ecr_fe_repo_url
  region          = var.region
  domain_name     = var.domain_name
  be_image_tag    = var.be_image_tag
  fe_image_tag    = var.fe_image_tag
}
