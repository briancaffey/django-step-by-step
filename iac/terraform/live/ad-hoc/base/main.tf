terraform {
  required_version = ">=1.3.6"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "4.48.0"
    }
  }

  backend "s3" {}
}

provider "aws" {
  region = var.region
  default_tags {
    tags = {
      env                  = terraform.workspace
      shared_resources_env = terraform.workspace
      shared_resource      = "true"
    }
  }
}

module "main" {
  source          = "git::https://github.com/briancaffey/terraform-aws-django.git//modules/ad-hoc/base"
  certificate_arn = var.certificate_arn
}
