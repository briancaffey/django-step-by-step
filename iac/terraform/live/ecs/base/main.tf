module "main" {
  # for local development
  # source          = "../../../../../../terraform-aws-django/modules/ecs/base"
  # for production use use git URL pointing to the module (recommended to use tags instead of main branch)
  source          = "git::https://github.com/briancaffey/terraform-aws-django.git//modules/ecs/base" # add ?ref=<branch> to use a branch
  certificate_arn = var.certificate_arn
  domain_name     = var.domain_name
}
