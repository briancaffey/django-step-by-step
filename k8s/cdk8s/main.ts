import { Construct } from 'constructs';
import { App, Chart } from 'cdk8s';
import * as k8s from './imports/k8s';

import { BackendApi } from './resources/api-deployment';
import { DatabaseResources } from './resources/database';

const NAMESPACE = process.env.NAMESPACE || 'app';

export class AppChart extends Chart {
  constructor(scope: Construct, id: string) {
    super(scope, id, {namespace: NAMESPACE});

    new k8s.KubeNamespace(this, 'AppNamespace', {
      metadata: {
        name: NAMESPACE,
        labels: {
          name: NAMESPACE
        }
      }
    });
    
    new BackendApi(this, 'BackendApi');
    new DatabaseResources(this, 'DatabaseResources');

  }
}

const app = new App();
new AppChart(app, 'cdk8s');
// synthesize the application to a single YAML file
app.synth();
