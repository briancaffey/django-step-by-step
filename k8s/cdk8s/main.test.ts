import {AppChart} from './main';
import {Testing} from 'cdk8s';

describe('Placeholder', () => {
  test('Empty', () => {
    const app = Testing.app();
    const chart = new AppChart(app, 'test-chart');
    const results = Testing.synth(chart)
    expect(results).toMatchSnapshot();
  });
});
