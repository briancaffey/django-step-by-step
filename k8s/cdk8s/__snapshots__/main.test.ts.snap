// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder Empty 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-chart-apideployment-c8d07176",
      "namespace": "my-app",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "api",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "api",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "bash",
                "-c",
                "python3 manage.py runserver_plus 0.0.0.0:8000",
              ],
              "env": Array [
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "backend.settings.development",
                },
              ],
              "imagePullPolicy": "",
              "name": "api",
              "ports": Array [
                Object {
                  "containerPort": 8000,
                  "name": "http",
                },
              ],
            },
          ],
          "initContainers": Array [
            Object {
              "command": Array [
                "bash",
                "-c",
                "python3 manage.py migrate",
              ],
              "env": Array [
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "backend.settings.development",
                },
              ],
              "image": "localhost:3200/backend:latest",
              "name": "migrate",
            },
          ],
          "restartPolicy": "Always",
        },
      },
    },
  },
]
`;
